pilot/utils/function_calling.py

```diff
@@ -0,0 +1,219 @@
+import json
+import re
+from typing import Literal, Optional, TypedDict, Callable
+
+JsonType = str | int | float | bool | None | list["JsonType"] | dict[str, "JsonType"]
+
+
+class FunctionParameters(TypedDict):
+    """Function parameters"""
+
+    type: Literal["object"]
+    properties: dict[str, JsonType]
+    required: Optional[list[str]]
+
+
+class FunctionType(TypedDict):
+    """Function type"""
+
+    name: str
+    description: Optional[str]
+    parameters: FunctionParameters
+
+
+class FunctionCall(TypedDict):
+    """Function call"""
+
+    name: str
+    parameters: str
+
+
+class FunctionCallSet(TypedDict):
+    definitions: list[FunctionType]
+    functions: dict[str, Callable]
+
+
+def add_function_calls_to_request(gpt_data, function_calls: FunctionCallSet | None):
+    if function_calls is None:
+        return
+
+    model: str = gpt_data['model']
+    is_llama = 'llama' in model or 'anthropic' in model

```

Comment:
why is it different for llama compared to all other llms, or do we have to add implementation for each llm individually?

Reply:
I need to rename this to `is_instruct`. Wrapping in `[INST]` seemed to have a negative affect for some models.